### PR TITLE
NPM adhering to both ingress and egress rules

### PR DESCRIPTION
**Pull request by vakalapa**
Currently NPM accepts a packet if it hits at least one rule in either ingress chain or egress chain. This behavior had an obvious flaw where if ingress allowed and egress blocked it, NPM would send the packets through.

This change will help evaluate both INGRESS and EGRESS rules before accepting/taking a decision on a packet. NPM will now MARK a packet for ingress/egress and RETURN the MARK'ed packet. Then packet will be accepted in the main chain after all the ingress and egress rules are processed.

Below is an example 

```
Chain AZURE-NPM (1 references)
 pkts bytes target     prot opt in     out     source               destination
  367 18860 AZURE-NPM-INGRESS-PORT  all  --  *      *       0.0.0.0/0            0.0.0.0/0
  367 18860 AZURE-NPM-EGRESS-PORT  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x3000 /* ACCEPT-on-INGRESS-and-EGRESS-mark-0x3000 */
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x2000 /* ACCEPT-on-INGRESS-mark-0x2000 */
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x1000 /* ACCEPT-on-EGRESS-mark-0x1000 */
  346 17600 AZURE-NPM-TARGET-SETS  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED

Chain AZURE-NPM-EGRESS-PORT (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 MARK       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set azure-npm-1078940524 dst tcp dpt:80 match-set azure-npm-2071429381 src /* ALLOW-ns-ns-name:network-policy-b-AND-TCP-PORT-80-FROM-ns-network-policy-2312 */ MARK or 0x1000
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x3000 /* RETURN-on-EGRESS-and-INGRESS-mark-0x3000 */
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x1000 /* RETURN-on-EGRESS-mark-0x1000 */
   21  1260 AZURE-NPM-TARGET-SETS  all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set azure-npm-2071429381 src /* ALLOW-ALL-FROM-ns-network-policy-2312-TO-JUMP-TO-AZURE-NPM-TARGET-SETS */

Chain AZURE-NPM-EGRESS-TO (0 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x3000 /* RETURN-on-EGRESS-and-INGRESS-mark-0x3000 */
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x1000 /* RETURN-on-EGRESS-mark-0x1000 */

Chain AZURE-NPM-INGRESS-FROM (0 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x2000 /* RETURN-on-INGRESS-mark-0x2000 */

Chain AZURE-NPM-INGRESS-PORT (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 MARK       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set azure-npm-2071429381 dst match-set azure-npm-1078940524 src tcp dpt:80 /* ALLOW-ns-ns-name:network-policy-b-AND-TCP-PORT-80-TO-ns-network-policy-2312 */ MARK set 0x2000
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x2000 /* RETURN-on-INGRESS-mark-0x2000 */
    0     0 AZURE-NPM-TARGET-SETS  all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set azure-npm-2071429381 dst /* ALLOW-ALL-TO-ns-network-policy-2312-TO-JUMP-TO-AZURE-NPM-TARGET-SETS */

Chain AZURE-NPM-TARGET-SETS (3 references)
 pkts bytes target     prot opt in     out     source               destination
   21  1260 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set azure-npm-2071429381 src /* DROP-ALL-FROM-ns-network-policy-2312 */
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set azure-npm-2071429381 dst /* DROP-ALL-TO-ns-network-policy-2312 */
```

### CODE REVIEW

**Comment by csfmomo** [comment]
Do we want to add a test case to hit this scenario? For example, ingress allow the certain packet however egress deny it, then the final result should be packet be denied. I think the test can be added in translatePolicy_test.go. Not sure whether our current test already covered it since I saw you also updated translatePolicy_test.go

File: npm/util/const.go [hunk 1]
```diff
@@ -75,6 +80,11 @@ const (
 	IptablesAzureIngressFromPodChain string = "AZURE-NPM-INGRESS-FROM-POD"
 	IptablesAzureEgressToNsChain     string = "AZURE-NPM-EGRESS-TO-NS"
 	IptablesAzureEgressToPodChain    string = "AZURE-NPM-EGRESS-TO-POD"
+	// Below are the ctmark NPM will use for different criteria
+	IptablesAzureIngressMarkHex string = "0x2000"
+	IptablesAzureEgressXMarkHex string = "0x1000/0x1000"
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply npm/util/const.go:85]: Correcting this, it is not a Connection Tracker mark, instead a mark field in skb.

**Comment by vakalapa** [comment]
@csfmomo Looks like the 2 policies are individually covered in our unit tests already. And with k8s e2e we will be covering them together, so I am skipping adding any new testcases. Let me know if you think otherwise

File: npm/util/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (neaggarwMS) [reply npm/util/const.go:85]: What is the difference between EgressX vs Egress. Can you update the comments and explain the significance of these margers

File: npm/translatePolicy.go [hunk 2]
```diff
@@ -224,7 +224,9 @@
 			entry.Specs = append(
```
Comment (neaggarwMS) [inline npm/translatePolicy.go:224]: Do we need to update the ItablesComment and append - MARK-0x100?

File: npm/util/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (neaggarwMS) [reply npm/util/const.go:85]: I guess you explained the difference in the Brownbag meeting, can you add comment here as well explaining what all these marks represent

**Review by neaggarwMS — CHANGES_REQUESTED** [review]

File: npm/translatePolicy.go (hunk 2 above)
Comment (vakalapa) [reply npm/translatePolicy.go:224]: I mean we can, but i am worried about the length of a given rule in some instances.

**Review by tamilmani1989 — COMMENTED** [review]
i didn't do detailed review as i don't have much context on this code. Looks like one big function. should break this function. May be separate PR.
Should add more comments wherever you are changing it. I see you added in few places. can you add in places wherever you could find

**Comment by vakalapa** [comment]
> i didn't do detailed review as i don't have much context on this code. Looks like one big function. should break this function. May be separate PR.
> Should add more comments wherever you are changing it. I see you added in few places. can you add in places wherever you could find

@tamilmani1989  totally agree with cutting down the large functions translateIngressPolicy and translateEgressPolicy. I will log it as a separate WI.

File: npm/iptm/iptm.go [hunk 3]
```diff
@@ -114,11 +114,65 @@
...
+		util.IptablesJumpFlag,
+		util.IptablesReturn,
+		util.IptablesModuleFlag,
+		util.IptablesMarkVerb,
+		util.IptablesMarkFlag,
+		util.IptablesAzureIngressMarkHex,
+		util.IptablesModuleFlag,
+		util.IptablesCommentModuleFlag,
+		util.IptablesCommentFlag,
+		"RETURN-on-INGRESS-mark-0x2000",
```
Comment (matmerr) [inline npm/iptm/iptm.go:129]: nit: can we use fmt.Sprintf("RETURN-on-INGRESS-mark-+v",IptablesAzureIngressMarkHex) to make sure the comments reflect the mark being added?

**Review by csfmomo — APPROVED** [review]

**Review by neaggarwMS — APPROVED** [review]

**Pull request merged**